code/part-one/tests/04-Mining.js

```diff
@@ -0,0 +1,140 @@
+    });
+  });
+
+  describe('MineableBlock', function() {
+    let previousHash = null;
+    let transactions = null;
+    let block = null;
+
+    beforeEach(function() {
+      const signer = signing.createPrivateKey();
+      const recipient = signing.getPublicKey(signing.createPrivateKey());
+      const amount = Math.ceil(Math.random() * 100);
+
+      transactions = [ new MineableTransaction(signer, recipient, amount) ];
+      previousHash = randomBytes(64).toString('hex');
+
+      block = new MineableBlock(transactions, previousHash);
+    });
+
+    it('should be instantiated without a hash', function() {
+      expect(!!block.hash).to.be.false;
+    });
+
+  });
+
+  describe('MineableChain', function() {
+    let blockchain = null;
+    let transaction = null;
+    let miner = null;
+
+    beforeEach(function() {
+      blockchain = new MineableChain();
+
+      const signer = signing.createPrivateKey();
+      const recipient = signing.getPublicKey(signing.createPrivateKey());
+      const amount = Math.ceil(Math.random() * 100);
+      transaction = new MineableTransaction(signer, recipient, amount);
+
+      miner = signing.createPrivateKey();
+      blockchain.addTransaction(transaction);
+      blockchain.mine(miner);
+    })
+
+    it('should have properties for difficulty and reward', function() {
+      expect(blockchain.difficulty).to.exist.and.be.a('number');
+      expect(blockchain.reward).to.exist.and.be.a('number');
+    });
+
+    it('should be able to mine new blocks', function() {
+      expect(blockchain.blocks).to.have.lengthOf(2);
+
+      const head = blockchain.getHeadBlock();
+      expect(head.transactions).to.deep.include(transaction);
+    });
+
+    it('should mine blocks with a valid zero-leading hash', function() {
+      const head = blockchain.getHeadBlock();
+      const zeros = Array(blockchain.difficulty.length).fill('0').join('');
+
+      expect(head.hash).to.be.a('string').and.not.be.empty;
+      expect(head.hash.slice(0, zeros.length)).to.equal(zeros);
+
+      const copy = new MineableBlock(head.transactions, head.previousHash);
+      copy.calculateHash(head.nonce);
+      expect(copy.hash).to.equal(head.hash);
+    });
+
+    it('should include a transaction rewarding the miner', function() {
+      const { transactions } = blockchain.getHeadBlock();
+      const reward = transactions.find(transaction => {
+        return transaction.recipient === signing.getPublicKey(miner)
+      });
+
+      expect(reward).to.not.be.undefined;
+      expect(reward.source).to.be.null;
+      expect(reward.amount).to.equal(blockchain.reward);
+    });
+
+    it('should clear out pending transactions after mining', function() {
+      // Add a transaction so we can mine a new block
+      blockchain.addTransaction(new MineableTransaction(
+        signing.createPrivateKey(),
+        signing.getPublicKey(signing.createPrivateKey()),
+        Math.ceil(Math.random() * 100)
+      ));
```

Comment:
Thoughts on testing to see that no new blocks were added btwn `blockchain.addTransaction(...)` and `blockchain.mine(...)` by inserting another `expect()` statement here? (eg, check to see if the original head block is still the current head block)

Reply:
I think that's a separate test ("should only add blocks when mining"), but I can add it.